### PR TITLE
nvme: Live Migration

### PR DESCRIPTION
**Pull request by NateThornton**
Implementation of TP 4159 PCIe Infrastructure for Live Migration plugin. Includes command support for Track Send, Migration Receive, Migration Send, and Controller Data Queue; Identify Controller LM related fields; Bash and ZSH completions.

Changes are isolated to the User Data Migration subset, with Track Memory functionality deferred to a future commit.

[TP4159 PCIe Infrastructure for Live Migration 2024.07.30 Ratified.pdf](https://github.com/user-attachments/files/17682852/TP4159.PCIe.Infrastructure.for.Live.Migration.2024.07.30.Ratified.pdf)

### CODE REVIEW

File: plugins/lm/lm-nvme.c [hunk 1]
```diff
@@ -0,0 +1,988 @@
...
+	DELETE_CONTROLLER_DATA_QUEUE = 1
+};
+
+static int lm_create_cdq(int argc, char **argv, struct command *command, struct plugin *plugin)
+{
+	const char *desc = "Create Controller Data Queue for controller of specific type and size";
+	const char *sz = "CDQ Size (in dwords)";
+	const char *cntlid = "Controller ID";
+	const char *qt = "Queue Type (default: 0 = User Data Migration Queue)";
+	const char *consent = "I consent this will not work and understand a CDQ cannot be mapped "
```
Comment (igaw) [inline plugins/lm/lm-nvme.c:91]: nice one!

File: plugins/lm/lm-nvme.c [hunk 2]
```diff
@@ -0,0 +1,988 @@
...
+	};
+
+	err = nvme_submit_admin_passthru(dev_fd(dev), &cmd, NULL);
+	if (!err)
+		printf("Migration Send (%s) Successful\n",
+		       ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+	else if (err > 0)
+		nvme_show_status(err);
+	else
+		nvme_show_error("ERROR: nvme_submit_admin_passthru() failed %s", strerror(errno));
```
Comment (igaw) [inline plugins/lm/lm-nvme.c:458]: I think this part should be in libnvme like the other helpers.

Comment (igaw) [reply plugins/lm/lm-nvme.c:458]: and with this in mind, the same comment is for the other places above. This is in the base spec, so we should also properly support it in the library and not have all the logic in the frontend.

File: plugins/lm/lm-nvme.c [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+struct nvme_io_submission_queue_data {
+	__u64 prp1;
+	__u16 qsize;
+	__u16 qid;
+	__u16 cqid;
+	__u16 attrs;
+	__u16 hp;
+	__u16 tp;
+	__u8  rsvd[4];
+};
```
Comment (igaw) [inline plugins/lm/lm-nvme.c:473]: these things really should be in libnvme

**Comment by NateThornton** [comment]
As we placed the new lm iocts in libnvme version 1.12, I think this change is still pending the next libnvme release.

Then one follow-up change to meson.build is required to grab that new release.

File: completions/_nvme [hunk 4]
```diff
@@ -617,6 +618,118 @@ _nvme () {
 				;;
 			esac
 			;;
+		(lm)
```
Comment (NateThornton) [inline completions/_nvme:621]: @NateThornton Self-reminder these need updating too

File: subprojects/libnvme.wrap [hunk 5]
```diff
@@ -1,6 +1,6 @@
 [wrap-git]
 url = https://github.com/linux-nvme/libnvme.git
-revision = 8efcec730603ffcc379ba0bf821e4c24ec87715b
+revision = 27ed889d0b1917919763816c97a49bb6e29d4547
```
Comment (igaw) [inline subprojects/libnvme.wrap:3]: btw, could split the libnvme build dependency into a separate patch and have it first? This makes any backporting way simpler.

File: nvme-print-json.c [hunk 6]
```diff
@@ -434,6 +434,16 @@ void json_nvme_id_ctrl(struct nvme_id_ctrl *ctrl,
...
 
+	obj_add_uint(r, "cmmrtd", le16_to_cpu(ctrl->cmmrtd));
+	obj_add_uint(r, "nmmrtd", le16_to_cpu(ctrl->nmmrtd));
+	obj_add_uint(r, "minmrtg", ctrl->minmrtg);
+	obj_add_uint(r, "maxmrtg", ctrl->maxmrtg);
+	obj_add_uint(r, "trattr", ctrl->trattr);
+	obj_add_uint(r, "mcumq", le16_to_cpu(ctrl->mcudmq));
+	obj_add_uint(r, "mcmr", le16_to_cpu(ctrl->mcmr));
+	obj_add_uint(r, "nmcmr", le16_to_cpu(ctrl->nmcmr));
+	obj_add_uint(r, "mcdqpc", le16_to_cpu(ctrl->mcdqpc));
```
Comment (igaw) [inline nvme-print-json.c:445]: the nvme-print*.c updates should be a separate patch.

File: completions/_nvme (hunk 4 above)
Comment (igaw) [reply completions/_nvme:621]: and while at, also split the tab completions changes into a separate patch. don't forget to update the docs too :)

File: plugins/lm/lm-nvme.c [hunk 7]
```diff
@@ -0,0 +1,894 @@
...
+		nvme_show_status(err);
+	else
+		printf("Migration Send (%s) Successful\n",
+		       ARGSTR(lm_migration_send_select_argstr, cfg.sel));
+
+
+	return err;
+}
+
+static void json_controller_state_data(struct nvme_lm_controller_state_data *data, size_t len)
```
Comment (igaw) [inline plugins/lm/lm-nvme.c:420]: So it might make sense to support builds without json-c dependencies for this new plugin. Most plugins depend on json-c to be available.  @ikegami-t started to split the plugins so they work also when json-c is missing (see the ocp plugin). I think it would be good to have this from the beginning as it is currently small code base. Though the LM is something in the base spec, so not unlikely to change or get udpated soonish :)

File: subprojects/libnvme.wrap (hunk 5 above)
Comment (NateThornton) [reply subprojects/libnvme.wrap:3]: Created #2616

File: nvme-print-json.c (hunk 6 above)
Comment (NateThornton) [reply nvme-print-json.c:445]: Glad you pointed this out because it's already included in #2607. Surprised I didn't get a merge conflict.

File: subprojects/libnvme.wrap (hunk 5 above)
Comment (igaw) [reply subprojects/libnvme.wrap:3]: Usually you just add a patch to a PR, no need for an extra PR just to update the dependency.

File: plugins/lm/lm-nvme.c (hunk 7 above)
Comment (NateThornton) [reply plugins/lm/lm-nvme.c]: Refactored following ocp plugin as an example.

**Comment by igaw** [comment]
The commit subject should be prefix, e.g. `plugins/lm: introduce live migration plugin`, `plugins/ls: add print files`, ...
Commit message shouldn't contain version information such as 'reverted ...'. After it gets merged no one cares about this things. If we really want to preserve the development history we would need to change a few things anyway.

Reverted nvme-print-stdout.c changes: still something in the first patch. As I said, I don't mind that nvme-print-stdout.c and nvme-print-json.c get updated, it  just deserves it's own patch.

Also why patch 2 removing stuff added previously. These patches should really be consisting out of new code being added and a tiny change in the meson.build file.

And the build is unhappy due to printf formatting issues.

**Comment by NateThornton** [comment]
I should have clarified my intent to squash the commits after they were individually reviewed. Everything now is in one commit and the build is passing.

I think the remaining warnings are acceptable, but if something needs to be done to clean them up LMK.

As we placed the new lm_* ioctls in release 12 of libnvme, does this merge need to wait until that is released?

**Comment by igaw** [comment]
I've dropped the changes in `nvme-print-stdout.c`. These were removing the output for the LM feature. 

Thanks!
